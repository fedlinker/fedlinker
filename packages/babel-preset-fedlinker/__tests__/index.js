const babel = require('@babel/core');

describe('babel-preset-fedlinker', () => {
  describe('basic', () => {
    const basicInput = `
      import React, { Component } from 'react'
      import propTypes from 'prop-types'
      class Button extends Component {
        render() {
          return (
            <button>Button</button>
          )
        }
      }
      Button.propTypes = { disabled: propTypes.bool }
      export default Button
    `;
    const basicOptions = {
      babelrc: false,
      presets: [[require('../index')]],
    };

    test('development', () => {
      process.env.BABEL_ENV = 'development';
      const { code } = babel.transformSync(basicInput, basicOptions);
      expect(code).toMatchSnapshot();
    });

    test('production', () => {
      process.env.BABEL_ENV = 'production';
      const { code } = babel.transformSync(basicInput, basicOptions);
      expect(code).toMatchSnapshot();
    });

    test('test', () => {
      process.env.BABEL_ENV = 'test';
      const { code } = babel.transformSync(basicInput, basicOptions);
      expect(code).toMatchSnapshot();
    });
  });

  describe('polyfills', () => {
    test('default polyfills', () => {
      process.env.BABEL_ENV = 'development';

      const options = {
        babelrc: false,
        presets: [[require('../index'), { entry: '/entry.js' }]],
        filename: '/entry.js',
      };

      const { code } = babel.transformSync(`const a = ''`, options);
      expect(code).toMatch('@babel/runtime-corejs3/core-js-stable/map.js');
      expect(code).toMatch('@babel/runtime-corejs3/core-js-stable/set.js');
      expect(code).toMatch(require.resolve('raf'));
    });

    test('with fetch polyfills', () => {
      process.env.BABEL_ENV = 'production';

      const options = {
        babelrc: false,
        presets: [[require('../index'), { entry: '/entry.js', fetchPolyfill: true }]],
        filename: '/entry.js',
      };

      const { code } = babel.transformSync(`const a = ''`, options);
      expect(code).toMatch('@babel/runtime-corejs3/core-js-stable/map.js');
      expect(code).toMatch('@babel/runtime-corejs3/core-js-stable/set.js');
      expect(code).toMatch(require.resolve('whatwg-fetch'));
      expect(code).toMatch(require.resolve('abortcontroller-polyfill'));
    });

    test('no default polyfills', () => {
      process.env.BABEL_ENV = 'production';
      const options = {
        babelrc: false,
        presets: [[require('../index'), { entry: '/entry.js', injectPolyfills: false }]],
        filename: '/entry.js',
      };

      const { code } = babel.transformSync(`const a = ''`, options);
      expect(code).not.toMatch('@babel/runtime-corejs3/core-js-stable/map.js');
      expect(code).not.toMatch('@babel/runtime-corejs3/core-js-stable/set.js');
      expect(code).not.toMatch(require.resolve('whatwg-fetch'));
      expect(code).not.toMatch(require.resolve('abortcontroller-polyfill'));
    });

    test('with user polyfills', () => {
      process.env.BABEL_ENV = 'development';

      const options = {
        babelrc: false,
        presets: [[require('../index'), { entry: '/entry.js', polyfills: ['module-name'] }]],
        filename: '/entry.js',
      };

      const { code } = babel.transformSync(`const a = ''`, options);
      expect(code).toMatch('@babel/runtime-corejs3/core-js-stable/map.js');
      expect(code).toMatch('@babel/runtime-corejs3/core-js-stable/set.js');
      expect(code).toMatch(require.resolve('raf'));
      expect(code).toMatch('module-name');
    });

    test('noly user polyfills', () => {
      process.env.BABEL_ENV = 'development';

      const options = {
        babelrc: false,
        presets: [
          [
            require('../index'),
            { entry: '/entry.js', injectPolyfills: false, polyfills: ['module-name'] },
          ],
        ],
        filename: '/entry.js',
      };

      const { code } = babel.transformSync(`const a = ''`, options);
      expect(code).not.toMatch('@babel/runtime-corejs3/core-js-stable/map.js');
      expect(code).not.toMatch('@babel/runtime-corejs3/core-js-stable/set.js');
      expect(code).not.toMatch(require.resolve('raf'));
      expect(code).toMatch('module-name');
    });
  });

  describe('static type checker', () => {
    test('flow', () => {
      process.env.BABEL_ENV = 'development';

      const input = `
        // @flow
        function square(n: number): number {
          return n * n;
        }
      `;

      const options = {
        babelrc: false,
        presets: [[require('../index'), { flow: true }]],
        filename: '/entry.js',
      };

      const { code } = babel.transformSync(input, options);
      expect(code).not.toMatch('square(n: number): number');
    });

    test('typescript', () => {
      process.env.BABEL_ENV = 'development';

      const input = `
        function identity<T>(arg: T): T {
          return arg;
        }
      `;

      const options = {
        babelrc: false,
        presets: [[require('../index'), { typescript: true }]],
        filename: '/entry.ts',
      };

      const { code } = babel.transformSync(input, options);
      expect(code).not.toMatch('identity<T>(arg: T): T');
    });
  });

  describe('proposals', () => {
    test('all', () => {
      process.env.BABEL_ENV = 'development';

      const input = `
// do-expressions
let a = do {
  if(x > 10) {
    'big';
  } else {
    'small';
  }
};

// nullish-coalescing-operator
var foo = object.foo ?? "default";

// function-bind
obj::func

// function-sent
function* generator() {
  console.log("Sent", function.sent);
  console.log("Yield", yield);
}

// logical-assignment-operators
a ||= b;

// numeric-separator
let budget = 1_000_000_000_000;

// optional-chaining
const baz = obj?.foo?.bar?.baz;

// partial-application
f(?, x);

// pipeline-operator
let result = "hello"
  |> doubleSay
  |> capitalize
  |> exclaim;

// private-methods
class Counter extends HTMLElement {
  #xValue = 0;
  get #x() { return this.#xValue; }
  set #x(value) {
    this.#xValue = value;
  }
  #clicked() {
    this.#x++;
  }
}

// throw-expressions
filename = throw new TypeError("Argument required")

// export-default-from
export v from 'mod';

// export-namespace-from
export * as ns from 'mod';
      `;

      const options = {
        babelrc: false,
        presets: [[require('../index'), { proposals: 'all' }]],
      };

      const { code } = babel.transformSync(input, options);
      expect(code).toMatchSnapshot();
    });
  });
});
