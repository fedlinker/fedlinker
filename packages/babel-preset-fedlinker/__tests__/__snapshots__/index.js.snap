// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`babel-preset-fedlinker basic development 1`] = `
"import _classCallCheck from \\"@babel/runtime-corejs3/helpers/esm/classCallCheck\\";
import _createClass from \\"@babel/runtime-corejs3/helpers/esm/createClass\\";
import _possibleConstructorReturn from \\"@babel/runtime-corejs3/helpers/esm/possibleConstructorReturn\\";
import _getPrototypeOf from \\"@babel/runtime-corejs3/helpers/esm/getPrototypeOf\\";
import _inherits from \\"@babel/runtime-corejs3/helpers/esm/inherits\\";
var _jsxFileName = \\"\\";
import React, { Component } from 'react';
import propTypes from 'prop-types';

var Button =
/*#__PURE__*/
function (_Component) {
  _inherits(Button, _Component);

  function Button() {
    _classCallCheck(this, Button);

    return _possibleConstructorReturn(this, _getPrototypeOf(Button).apply(this, arguments));
  }

  _createClass(Button, [{
    key: \\"render\\",
    value: function render() {
      return React.createElement(\\"button\\", {
        __source: {
          fileName: _jsxFileName,
          lineNumber: 7
        },
        __self: this
      }, \\"Button\\");
    }
  }]);

  return Button;
}(Component);

Button.propTypes = {
  disabled: propTypes.bool
};
export default Button;"
`;

exports[`babel-preset-fedlinker basic production 1`] = `
"import _classCallCheck from \\"@babel/runtime-corejs3/helpers/esm/classCallCheck\\";
import _createClass from \\"@babel/runtime-corejs3/helpers/esm/createClass\\";
import _possibleConstructorReturn from \\"@babel/runtime-corejs3/helpers/esm/possibleConstructorReturn\\";
import _getPrototypeOf from \\"@babel/runtime-corejs3/helpers/esm/getPrototypeOf\\";
import _inherits from \\"@babel/runtime-corejs3/helpers/esm/inherits\\";
import React, { Component } from 'react';

var _ref =
/*#__PURE__*/
React.createElement(\\"button\\", null, \\"Button\\");

var Button =
/*#__PURE__*/
function (_Component) {
  _inherits(Button, _Component);

  function Button() {
    _classCallCheck(this, Button);

    return _possibleConstructorReturn(this, _getPrototypeOf(Button).apply(this, arguments));
  }

  _createClass(Button, [{
    key: \\"render\\",
    value: function render() {
      return _ref;
    }
  }]);

  return Button;
}(Component);

export default Button;"
`;

exports[`babel-preset-fedlinker basic test 1`] = `
"\\"use strict\\";

var _interopRequireWildcard = require(\\"@babel/runtime-corejs3/helpers/interopRequireWildcard\\");

var _interopRequireDefault = require(\\"@babel/runtime-corejs3/helpers/interopRequireDefault\\");

var _Object$defineProperty = require(\\"@babel/runtime-corejs3/core-js-stable/object/define-property\\");

_Object$defineProperty(exports, \\"__esModule\\", {
  value: true
});

exports.default = void 0;

var _react = _interopRequireWildcard(require(\\"react\\"));

var _propTypes = _interopRequireDefault(require(\\"prop-types\\"));

var _jsxFileName = \\"\\";

class Button extends _react.Component {
  render() {
    return _react.default.createElement(\\"button\\", {
      __source: {
        fileName: _jsxFileName,
        lineNumber: 7
      },
      __self: this
    }, \\"Button\\");
  }

}

Button.propTypes = {
  disabled: _propTypes.default.bool
};
var _default = Button;
exports.default = _default;"
`;

exports[`babel-preset-fedlinker proposals all 1`] = `
"import _regeneratorRuntime from \\"@babel/runtime-corejs3/regenerator\\";
import _Object$defineProperty from \\"@babel/runtime-corejs3/core-js-stable/object/define-property\\";
import _concatInstanceProperty from \\"@babel/runtime-corejs3/core-js-stable/instance/concat\\";
import _classCallCheck from \\"@babel/runtime-corejs3/helpers/esm/classCallCheck\\";
import _possibleConstructorReturn from \\"@babel/runtime-corejs3/helpers/esm/possibleConstructorReturn\\";
import _getPrototypeOf from \\"@babel/runtime-corejs3/helpers/esm/getPrototypeOf\\";
import _assertThisInitialized from \\"@babel/runtime-corejs3/helpers/esm/assertThisInitialized\\";
import _inherits from \\"@babel/runtime-corejs3/helpers/esm/inherits\\";
import _wrapNativeSuper from \\"@babel/runtime-corejs3/helpers/esm/wrapNativeSuper\\";
import _classPrivateFieldLooseBase from \\"@babel/runtime-corejs3/helpers/esm/classPrivateFieldLooseBase\\";
import _classPrivateFieldLooseKey from \\"@babel/runtime-corejs3/helpers/esm/classPrivateFieldLooseKey\\";
import _skipFirstGeneratorNext from \\"@babel/runtime-corejs3/helpers/esm/skipFirstGeneratorNext\\";
import \\"regenerator-runtime/runtime\\";
import _bindInstanceProperty from \\"@babel/runtime-corejs3/core-js-stable/instance/bind\\";

var _object$foo, _context, _context2, _obj, _obj$foo, _obj$foo$bar, _x, _f, _ref, _ref2, _hello;

// do-expressions
var a = x > 10 ? 'big' : 'small'; // nullish-coalescing-operator

var foo = (_object$foo = object.foo) !== null && _object$foo !== void 0 ? _object$foo : \\"default\\"; // function-bind

_bindInstanceProperty(_context2 = (_context = obj, func)).call(_context2, _context); // function-sent


function generator() {
  return _generator.apply(this, arguments);
} // logical-assignment-operators


function _generator() {
  _generator = _skipFirstGeneratorNext(
  /*#__PURE__*/
  _regeneratorRuntime.mark(function _callee() {
    var _functionSent;

    return _regeneratorRuntime.wrap(function _callee$(_context4) {
      while (1) {
        switch (_context4.prev = _context4.next) {
          case 0:
            _context4.next = 2;
            return;

          case 2:
            _functionSent = _context4.sent;
            console.log(\\"Sent\\", _functionSent);
            _context4.t0 = console;
            _context4.next = 7;
            return;

          case 7:
            _context4.t1 = _functionSent = _context4.sent;

            _context4.t0.log.call(_context4.t0, \\"Yield\\", _context4.t1);

          case 9:
          case \\"end\\":
            return _context4.stop();
        }
      }
    }, _callee);
  }));
  return _generator.apply(this, arguments);
}

a || (a = b); // numeric-separator

var budget = 1000000000000; // optional-chaining

var baz = (_obj = obj) == null ? void 0 : (_obj$foo = _obj.foo) == null ? void 0 : (_obj$foo$bar = _obj$foo.bar) == null ? void 0 : _obj$foo$bar.baz; // partial-application

_f = f, _x = x, function f(_argPlaceholder) {
  return _f(_argPlaceholder, _x);
}; // pipeline-operator

var result = (_ref = (_ref2 = (_hello = \\"hello\\", doubleSay(_hello)), capitalize(_ref2)), exclaim(_ref)); // private-methods

var Counter =
/*#__PURE__*/
function (_HTMLElement) {
  _inherits(Counter, _HTMLElement);

  function Counter() {
    var _getPrototypeOf2, _context3;

    var _this;

    _classCallCheck(this, Counter);

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = _possibleConstructorReturn(this, (_getPrototypeOf2 = _getPrototypeOf(Counter)).call.apply(_getPrototypeOf2, _concatInstanceProperty(_context3 = [this]).call(_context3, args)));

    _Object$defineProperty(_assertThisInitialized(_this), _clicked, {
      value: _clicked2
    });

    _Object$defineProperty(_assertThisInitialized(_this), _x2, {
      get: _get_x,
      set: _set_x
    });

    _Object$defineProperty(_assertThisInitialized(_this), _xValue, {
      writable: true,
      value: 0
    });

    return _this;
  }

  return Counter;
}(_wrapNativeSuper(HTMLElement)); // throw-expressions


var _xValue = _classPrivateFieldLooseKey(\\"xValue\\");

var _x2 = _classPrivateFieldLooseKey(\\"x\\");

var _clicked = _classPrivateFieldLooseKey(\\"clicked\\");

var _get_x = function _get_x() {
  return _classPrivateFieldLooseBase(this, _xValue)[_xValue];
};

var _set_x = function _set_x(value) {
  _classPrivateFieldLooseBase(this, _xValue)[_xValue] = value;
};

var _clicked2 = function _clicked2() {
  _classPrivateFieldLooseBase(this, _x2)[_x2]++;
};

filename = function (e) {
  throw e;
}(new TypeError(\\"Argument required\\")); // export-default-from


import _v from 'mod';
export { _v as v }; // export-namespace-from

import * as _ns from 'mod';
export { _ns as ns };"
`;
